Add wxSpinCtrl::GetTextValue()

vadz opened the pull request:
This allows to retrieve the current contents of the text entry part of
wxSpinCtrl.

See [#19140](https://trac.wxwidgets.org/ticket/19140)

Commit:
Add wxSpinCtrl::GetTextValue()

This allows to retrieve the current contents of the text entry part of
wxSpinCtrl.

For now provide a stub in the base class, will be made pure virtual
later.

Co-Authored-By: Vadim Zeitlin <vadim@wxwidgets.org>

vadz committed:
Also implement wxSpinCtrl::GetTextValue() for wxGTK1

vadz committed:
Implement wxSpinCtrl::GetTextValue() for wxQt too

vadz committed:
Make wxSpinCtrl::GetTextValue() pure virtual

It is now implemented in all ports.

vadz committed:
Always handle "input" and "output" signals in wxGTK wxSpinCtrl

No real changes yet, just refactor the code to always connect these
signals handlers and not just when using non-decimal base.

Add wxSpinCtrl::GTKInput() and GTKOutput() virtual methods to handle
these signals appropriately depending on the type of the control.

This will allow further customizing conversion to/from string in the
upcoming commit and will also make it possible to handle UI locale
different from the C locale in the future.

vadz committed:
Get rid of CppUnit boilerplate in wxSpinCtrlDouble unit test

This was already done for wxSpinCtrl unit tests in a4928c0fde (Use Catch
in wxSpinCtrl unit tests, 2020-07-12), do it for wxSpinCtrlDouble too
now.

Commit:
Allow clearing spin controls in the widgets sample

Calling wxSpinCtrl::SetValue("") is always valid.

See #19140.

vadz committed:
Replace dynamic cast with a virtual GTKValueChanged() function

No real changes, but using a virtual function is simpler and safer than
using wxIsKindOf().

vadz committed:
Allow using a different text value in wxGTK wxSpinCtrl

Make it possible to use a string value different from the numeric value,
as wxMSW version allows this and some existing code depends on it.

Closes #19140.

vadz committed:
Explicitly convert from UTF-8 in wxSpinCtrl::GetTextValue()

Use UTF-8 encoding explicitly rather than relying on it being the
default.

vadz committed:
Simplify event sending code by using GetTextValue()

Just use this function now that it's available rather than calling
gtk_entry_get_text() (and sometimes forgetting to specify UTF-8 encoding
explicitly).

vadz committed:
Document wxSpinCtrl::SetValue("") behaviour

Mention that this is explicitly allowed now that it works.

Co-Authored-By: Dummy <dummy@trac.wxwidgets.org>

vadz commented:
@paulcor I suspect you might not like this very much, but please let me know if you really object to merging this strenuously. I'd like to merge it relatively soon because I have more changes to `wxSpinCtrl` code that would conflict (or even depend on) these ones, so please let me know if you do. TIA!

paulcor commented on src/gtk/spinctrl.cpp:
```diff
@@ -523,4 +661,28 @@ void wxSpinCtrlDouble::SetDigits(unsigned digits)
+
+bool wxSpinCtrlDouble::GTKOutput(wxString* text) const
+{
+    if ( wxSpinCtrlGTKBase::GTKOutput(text) )
```
There doesn't seem to be much point in making `GTKOutput()` pure virtual when one of the overrides just calls the base. Unless you're planning something more here later. Could be done more simply as just
```suggestion
    return wxSpinCtrlGTKBase::GTKOutput(text);
```

vadz replied:
Sorry, I should have mentioned it, but I do plan to do more in it, the "new UI locale API" thread on wx-dev -- basically I'm going to convert between strings and doubles manually here soon (BTW, if you have any thoughts on feasibility of the solution proposed in this thread for wxGTK, I'd love to hear them).

vadz commented on src/gtk/spinctrl.cpp:
```diff
@@ -67,16 +52,57 @@ gtk_value_changed(GtkSpinButton* spinbutton, wxSpinCtrlGTKBase* win)
+    return FALSE;
+}
+
+static gint
```
Oops, indeed, thanks. I've fixed the return type of the "input" handler to correctly use `gint` rather than `gboolean` but didn't realize that "output" was symmetrically wrong in the other direction...

vadz committed:
Fix return type of GtkSpinButton "output" signal handler

This doesn't seem to change anything in practice, but still use the
correct return type.

Co-Authored-By: Paul Cornett <paulcor@users.noreply.github.com>

vadz committed:
Move wxEVT_TEXT generation to wxSpinCtrlGTKBase::GTKTextChanged()

This is more consistent with GTKValueChanged() and allows making the
change in the upcoming commit.

No real changes yet, this is just a refactoring.

vadz committed:
Reset text override if the text is manually modified

Otherwise the newly entered text wouldn't be taken into account, as it
would still be overridden.

vadz committed:
Fix resetting text override when manually entering the text

The entered text wasn't taken into account, as the override was still
used when its numeric value was retrieved by GtkSpinButton using our
"input" handler, so reset the override now as soon as we get "changed"
signal.
